ompi/info/info_memkind.c

```diff
@@ -0,0 +1,528 @@
+static ompi_memkind_t *ompi_info_memkind_available;
+static int ompi_info_memkind_num_available = 0;
+
+#if 0
+static void ompi_info_memkind_dump (const char *var_name, int num_memkinds, ompi_memkind_t *memkinds)
+{
+    for (int i = 0; i < num_memkinds; i++) {
+        printf("[%d] %s memkind[%d].name: %s ", getpid(), var_name, i, memkinds[i].im_name);
+        if (memkinds[i].im_num_restrictors > 0) {
+            printf("restrictors: ");
+            for (int j = 0; j < memkinds[i].im_num_restrictors; j++) {
+                printf("%c %s", (j == 0 ? ' ': ','), memkinds[i].im_restrictors[j]);
+            }
+        } else {
+            printf("num restrictors = 0");
+        }
+        printf(" no_restrictors = %s", memkinds[i].im_no_restrictors ? "true" : "false");
+        printf("\n");
+    }
+}
+#endif
+
+static void ompi_info_memkind_extract (const char* memkind_str, int *num_memkinds, ompi_memkind_t **memkinds)
+{
+    /* The memkind string is a comma-separated list of memkinds, which can have two forms:
+    **   - standalone memkind type, which implies that all restrictors of the memkind are requested
+    **     (or looking at it the other way around, no restrictions are imposed on that memory kind)
+    **   - memkind:restrictor
+    ** The same memkind type can appear multiple times, e.g.
+    **    memkind_a:restrictor_1,memkind_a:restrictor_2;
+    */
+
+    /* Separate requested_str into an array of individual entries */
+    char **memkind_combos = opal_argv_split(memkind_str, ',');
+    int max_num_memkinds = opal_argv_count(memkind_combos);
+
+    ompi_memkind_t *memkind_arr = NULL;
+    memkind_arr = (ompi_memkind_t *) malloc(max_num_memkinds * sizeof(ompi_memkind_t));
+    if (NULL == memkind_arr) {
+        goto err_exit;
+    }
+    for (int i = 0; i < max_num_memkinds; i++) {
+        memkind_arr[i].im_num_restrictors = 0;
+        memkind_arr[i].im_no_restrictors = true;
+    }
+
+    int iter = 0;
+    char *m = memkind_combos[iter];
+    int current_max = 0;
+    while (m != NULL) {
+        bool name_found = false;
+        char **tmp_str = opal_argv_split (m, ':');
+        int pos;
+
+        // Try to remove duplicates of the exact same name:restrictor appearance
+        for (int i = 0; i < current_max; i++) {
+            if (!strncmp(tmp_str[0], memkind_arr[i].im_name, strlen(tmp_str[0]))) {
+                name_found = true;
+                pos = i;
+                break;
+            }
+        }
+
+        if (name_found)  {
+            // check whether restrictor matches (if any present at either m
+            // or at memkind_arr[pos] )
+            if (NULL != tmp_str[1] ) {
+                if (memkind_arr[pos].im_num_restrictors > 0) {
+                    if (!strncmp(tmp_str[1], memkind_arr[pos].im_restrictors[0], strlen(tmp_str[1]))) {
+                        // We have seen this exact name:restrictor combination already                    
+                        continue;
```

Comment:
I'm a little confused about this: wouldn't this lead to an infinite loop because we don't advance `m` here?

Reply:
yes, good catch, working on resolving it. I did not have a testcase that triggered this code path.

Reply:
this should be resolved and fixed.